hudi-common/src/main/java/org/apache/hudi/common/util/TableInternalSchemaUtils.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.util;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hudi.common.model.HoodieCommitMetadata;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieTimeline;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.internal.schema.InternalSchema;
+import org.apache.hudi.internal.schema.utils.InternalSchemaParser;
+
+import java.util.TreeMap;
+
+/**
+ * helper class.
+ * use static cache to cache historySchema info to prevent fetching from hdfs every time.
+ * all task belong to the same executor/container will share this cache.
+ *
+ */
+public class TableInternalSchemaUtils {
+  private static final Cache<String, TreeMap<Long, InternalSchema>>
+      HISTORICAL_SCHEMA_CACHE = Caffeine.newBuilder().maximumSize(1000).weakValues().build();
+
+  public static InternalSchema searchSchemaAndCache(long versionID, String tablePath, Configuration hadoopConf) {
+    TreeMap<Long, InternalSchema> historicalSchemas = HISTORICAL_SCHEMA_CACHE.getIfPresent(tablePath);
+    if (historicalSchemas == null || InternalSchemaParser.searchSchema(versionID, historicalSchemas) == null) {
+      historicalSchemas = getHistoricalSchemas(tablePath, hadoopConf);
+      HISTORICAL_SCHEMA_CACHE.put(tablePath, historicalSchemas);
+    } else {
+      long maxVersionId = historicalSchemas.keySet().stream().max(Long::compareTo).get();
+      if (versionID > maxVersionId) {
+        historicalSchemas = getHistoricalSchemas(tablePath, hadoopConf);
+        HISTORICAL_SCHEMA_CACHE.put(tablePath, historicalSchemas);
+      }
+    }
+    return InternalSchemaParser.searchSchema(versionID, historicalSchemas);
+  }
+
+  private static TreeMap<Long, InternalSchema> getHistoricalSchemas(String tablePath, Configuration hadoopConf) {
+    TreeMap<Long, InternalSchema> result = new TreeMap<>();
```

Comment:
What about thread-safety? If there are concurrent readers, then should construction of TreeMap be synchronized?